cdap-common/src/test/java/co/cask/cdap/common/ssh/SSHSessionTest.java

```diff
@@ -169,6 +189,120 @@ public void testSsh() throws Exception {
+        }
+
+        @Override
+        public void finished() {
+          finished.set(true);
+        }
+      };
+
+
+      SSHConfig sshConfig = getSSHConfig();
+
+      // Creates a SSH session.
+      try (SSHSession session = new DefaultSSHSession(sshConfig)) {
+        InetSocketAddress bindAddress = echoServer.getBindAddress();
+
+        // Creates local port forward and send data to the echo server through that forwarding channel
+        try (PortForwarding portForwarding = session.createLocalPortForward(bindAddress.getHostName(),
+                                                                            bindAddress.getPort(),
+                                                                            12345, dataConsumer)) {
+          List<String> messages = new ArrayList<>();
+          for (int i = 0; i < 10; i++) {
+            String msg = "Testing" + i;
+            portForwarding.write(StandardCharsets.UTF_8.encode(msg));
+            portForwarding.write(StandardCharsets.UTF_8.encode("\n"));
+            messages.add(msg);
+          }
+          portForwarding.flush();
+
+          Iterable<String> splits = Splitter.on("\n").omitEmptyStrings().split(received);
+          Assert.assertEquals(messages, StreamSupport.stream(splits.spliterator(), false).collect(Collectors.toList()));
+        }
+
+        // After closing the port forwarding, the data consumer should have finished.
+        Assert.assertTrue(finished.get());
+      }
+
+    } finally {
+      echoServer.stopAndWait();
+    }
+  }
+
+  @Test
+  public void testForwardingOnSessionClose() throws Exception {
+    EchoServer echoServer = new EchoServer();
+
+    echoServer.startAndWait();
+    try {
+      SSHConfig sshConfig = getSSHConfig();
+      AtomicBoolean finished = new AtomicBoolean(false);
+      PortForwarding portForwarding;
+
+      // Creates a SSH session
+      try (SSHSession session = new DefaultSSHSession(sshConfig)) {
+        InetSocketAddress bindAddress = echoServer.getBindAddress();
+
+        // Creates a port forwarding and send some data
+        BlockingQueue<String> received = new LinkedBlockingQueue<>();
+        portForwarding = session.createLocalPortForward(bindAddress.getHostName(), bindAddress.getPort(),
+                                                        12345, new PortForwarding.DataConsumer() {
+          @Override
+          public void received(ByteBuffer buffer) {
+            received.add(StandardCharsets.UTF_8.decode(buffer).toString());
+          }
+
+          @Override
+          public void finished() {
+            finished.set(true);
+          }
+        });
+
+        portForwarding.write(StandardCharsets.UTF_8.encode("Testing"));
+        portForwarding.flush();
+
+        Assert.assertEquals("Testing", received.poll(5, TimeUnit.SECONDS));
+      }
+
+      // After closing of the SSH session, the port forwarding should be closed as well
+      Assert.assertTrue(finished.get());
+
```

Comment:
seems like we should also test portForwarding.isOpen() in these test cases

Reply:
It is already tested by writing data out to the Echo server through the port forwarding and receiving the echo back.